### PR TITLE
Preload / Prewarm custom models via URL on startup

### PR DESCRIPTION
**Pull request by vicilliar**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

* **What is the current behavior?** (You can also link to an open issue here)
Currently, the env var only accepts strings.

* **What is the new behavior (if this is a feature change)?**
This allows the env var `MARQO_MODELS_TO_PRELOAD` to accept custom models as dicts with `model` and `model_properties`. A sample declaration would look like this:

```
export MARQO_MODELS_TO_PRELOAD='[
        "ViT-L/14",
         {
            "model": "generic-clip-test-model-2",
            "model_properties": {
                "name": "ViT-B/32",
                "dimensions": 512,
                "type": "clip",
                "url": "https://openaipublic.azureedge.net/clip/models/40d365715913c9da98579312b702a82c18be219cc2a73407c4526f58eba950af/ViT-B-32.pt"
            }
        }
    ]'
```

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
Manual testing. Application tests will be made once the env var customization feature in Marqo API tests finishes

* **Related documentation changes** (link commit/PR here)
To follow

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by vicilliar** [comment]
Manual testing matrix (OPEN CLIP and CLIP only so far)

Please let me know if there are other important edge cases to test
image

**Comment by vicilliar** [comment]
Models used for said manual testing:
OPEN CLIP
```
{
            "model": "random-open-clip-1",
            "model_properties": {
                "name": "ViT-B-32-quickgelu",
                "dimensions": 512,
                "url": "https://github.com/mlfoundations/open_clip/releases/download/v0.2-weights/vit_b_32-quickgelu-laion400m_avg-8a00ab3c.pt",
                "type": "open_clip"
            }
        }
```

CLIP
```
{
            "model": "generic-clip-test-model-2",
            "model_properties": {
            	"name": "ViT-B/32",
                "dimensions": 512,
                "url": "https://openaipublic.azureedge.net/clip/models/40d365715913c9da98579312b702a82c18be219cc2a73407c4526f58eba950af/ViT-B-32.pt",
                "type": "clip"
            }
        }
```

File: src/marqo/tensor_search/on_start_script.py
Comment (pandu-k) [inline src/marqo/tensor_search/on_start_script.py]: As a general rule we should avoid nested functions if we can. My advice is make it a top level function, and add a bunch of unit tests for it. The unit tests should use mocking to isolate the logic being tested to just the function being tested.  

The unit tests, even if seemingly trivial, will make the system more stable as the project grows. See [here](https://github.com/marqo-ai/marqo/blob/mainline/tests/s2_inference/test_clip_utils.py#L168) for examples.

**Comment by pandu-k** [comment]
> 

Thanks for the testing matrix!

Is the no "name" behaviour the same as when creating an index with this model normally? What is the expected behaviour @wanliAlex ?

**Comment by wanliAlex** [comment]
> > 
> 
> Thanks for the testing matrix!
> 
> Is the no "name" behaviour the same as when creating an index with this model normally? What is the expected behaviour @wanliAlex ?



> > 
> 
> Thanks for the testing matrix!
> 
> Is the no "name" behaviour the same as when creating an index with this model normally? What is the expected behaviour @wanliAlex ?

Yes, because `"name"` is not required for loading clip models, but we make it a required filed in model properties.

**Comment by wanliAlex** [comment]
Do we want to hold this PR for a while until my custom HF models are merged? Do we want to test the custom HF models prewarming?

**Comment by wanliAlex** [comment]
The name check is [here](https://github.com/marqo-ai/marqo/blob/7acd2494e30326ff6451a39a6cfd9d75ffcc327f/src/marqo/s2_inference/s2_inference.py#L205-L215)

But note that the `"name"` is not used for clip models if a URL is provided.

**Comment by vicilliar** [comment]
> Do we want to hold this PR for a while until my custom HF models are merged? Do we want to test the custom HF models prewarming?

I'm under the impression this URL pre-warming feature is high priority, so it can be merged asap, but I'll need confirmation from @pandu-k 

If ever, do we see the format for HF/S3 models being much different? Or are they also in the form `model` and `model_properties`?

**Comment by wanliAlex** [comment]
> > Do we want to hold this PR for a while until my custom HF models are merged? Do we want to test the custom HF models prewarming?
> 
> I'm under the impression this URL pre-warming feature is high priority, so it can be merged asap, but I'll need confirmation from @pandu-k
> 
> If ever, do we see the format for HF/S3 models being much different? Or are they also in the form `model` and `model_properties`?

Yeah they are similar. Feel free to merge if it's urgent as this is for production use.

**Comment by pandu-k** [comment]
Let's get this merged in first, so we can spin up a testing tag

**Comment by wanliAlex** [comment]
The PR looks good to me. Just to confirm, did you do the manual test on a dockerised marqo? @vicilliar

**Comment by vicilliar** [comment]
> 

Ah good idea, i have not yet. Will do so

**Comment by vicilliar** [comment]
I did the tests on a local running marqo

**Comment by wanliAlex** [comment]
> I did the tests on a local running marqo

Can you try to run several manual tests on a dockerised marqo and upload some screenshots? Would this be enough to test the production environment? @pandu-k

**Comment by pandu-k** [comment]
> 

That should be pretty good. Try expected failure and expected successful cases. Ensure that search/add docs works as expected. The testing matrix was super useful

File: tests/tensor_search/test_on_start_script.py
Comment (wanliAlex) [inline tests/tensor_search/test_on_start_script.py:155]: can we have one more test to ensure the prewarmed model will not need to be download/load again?

**Review by wanliAlex — CHANGES_REQUESTED** [review]

File: tests/tensor_search/test_on_start_script.py
Comment (wanliAlex) [reply tests/tensor_search/test_on_start_script.py:155]: > can we have one more test to ensure the prewarmed model will not need to be download/load again?

For both search and add_document.

**Comment by pandu-k** [comment]
- unit tests: https://github.com/marqo-ai/marqo/actions/runs/4988240487 [passed]
- API tests: https://github.com/marqo-ai/marqo/actions/runs/4988241204 [passed]

File: tests/tensor_search/test_on_start_script.py
Comment (vicilliar) [reply tests/tensor_search/test_on_start_script.py:155]: this we can save for the API end-to-end test

**Comment by vicilliar** [comment]
failing 1 unit test on local:
This is probably just an existing problem, unit tests through GH actions work just fine
```
marqo.errors.BadRequestError: BadRequestError: Problem vectorising query. Reason: Unable to load model=ViT-B/16 on device=cpu with normalization=True. If you are trying to load a custom model, please check that model_properties={'name': 'ViT-B/16', 'dimensions': 512, 'notes': 'CLIP ViT-B/16', 'type': 'clip'} is correct and Marqo has access to the weights file.
```

```
============================================== short test summary info ==============================================
FAILED tests/tensor_search/test_add_documents.py::TestAddDocuments::test_mappings_arent_updated_images - marqo.errors.BadRequestError: BadRequestError: Problem vectorising query. Reason: Unable to load model=ViT-B/16 ...
```

**Comment by vicilliar** [comment]
Dockerized manual marqo testing:
works as expected (open clip).
1. Startup loads the models properly (hf model and a custom open clip model)
image

2. Upon indexing, no preloading is done (expected)
image

3. Search works fine, here is a sample query result for q="vegetable"
```
{'hits': [{'field_1': 'cabbage', '_id': '3', '_highlights': {'field_1': 'cabbage'}, '_score': 0.87822014}, {'field_1': 'red', '_id': '2', '_highlights': {'field_1': 'red'}, '_score': 0.8071192}, {'field_1': 'hello', '_id': '1', '_highlights': {'field_1': 'hello'}, '_score': 0.80202985}], 'query': 'vegetable', 'limit': 10, 'offset': 0, 'processingTimeMs': 284}
```

Here is the indexing/searching script used
```
import marqo
mq = marqo.Client()
settings = {
    "index_defaults": {
        "treat_urls_and_pointers_as_images": True,
        "model": 'generic-clip-test-model-1',
        "model_properties": {
            "name": "ViT-B-32-quickgelu",
                "dimensions": 512,
                "url": "https://github.com/mlfoundations/open_clip/releases/download/v0.2-weights/vit_b_32-quickgelu-laion400m_avg-8a00ab3c.pt",
                "type": "open_clip"
            },
        "normalize_embeddings": True,
    },
}
try:
    mq.delete_index("my-own-clip")
except:
    pass

response = mq.create_index("my-own-clip", settings_dict=settings)

mq.index("my-own-clip").add_documents(
    [
        {"_id": "1", "field_1": "hello"}, 
        {"_id": "2", "field_1": "red"}, 
        {"_id": "3", "field_1": "cabbage"}, 
    ]
)
res = mq.index("my-own-clip").search("vegetable")
print("search results are: ")
print(res)
```

**Comment by vicilliar** [comment]
Failure cases are also as expected:

Malformed JSON result:
image

Missing `model` result:
image


Missing `model_properties` result:
image

Also there are some mistakes in the error message, so I'm pushing a small change, just changing the message to this:
image

**Review by wanliAlex — APPROVED** [review]

**Pull request merged**
